ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ClasspathComputer.java

```diff
@@ -267,17 +268,23 @@ private static void addLibraryEntry(IPluginLibrary library, Map<String, IPath> s
 
 		IPackageFragmentRoot root = context.javaProject.getPackageFragmentRoot(jarFile);
 		if (root.exists() && root.getKind() == IPackageFragmentRoot.K_BINARY) {
-			IClasspathEntry oldEntry = root.getRawClasspathEntry();
-			// If we have the same binary root but new or different source, we
-			// should recreate the entry. That is when the source attachment:
-			// - is not defined: the default could be available now, or
-			// - is overridden with a different value.
-			if ((sourceAttachment == null && oldEntry.getSourceAttachmentPath() != null)
-					|| (sourceAttachment != null && sourceAttachment.equals(oldEntry.getSourceAttachmentPath()))) {
-				context.reloaded.add(oldEntry);
-				return;
+			try {
+				IClasspathEntry oldEntry = root.getRawClasspathEntry();
+				// If we have the same binary root but new or different source, we
+				// should recreate the entry. That is when the source attachment:
+				// - is not defined: the default could be available now, or
+				// - is overridden with a different value.
+				if ((sourceAttachment == null && oldEntry.getSourceAttachmentPath() != null)
+						|| (sourceAttachment != null && sourceAttachment.equals(oldEntry.getSourceAttachmentPath()))) {
+					context.reloaded.add(oldEntry);
+					return;
+				}
+				isExported = oldEntry.isExported();
+			} catch (JavaModelException e) {
+				if (e.getStatus().getCode() != IJavaModelStatusConstants.ELEMENT_NOT_ON_CLASSPATH) {
```

Comment:
So we see cases where classpath entry is not on the classpath.`IJavaModelStatusConstants.ELEMENT_NOT_ON_CLASSPATH` because someone modified classpath during this method execution?

What about to throw dedicated exception so that the caller `addSourceAndLibraries() `could re-iterate over the classpath again?

Reply:
It sounds like we need a real fix here for `PerProjectInfo.rootPathToRawEntries`?
Ideally you could update PR description what happened and how the change is supposed to fix that. Right now I'm clueless.

Reply:
Requiring UpdateClasspathJob to deal with the old state to potentially change _while_ the job is running somehow fails to make good feelings.

Reply:
I updated the PR description https://github.com/eclipse-pde/eclipse.pde/pull/1601#issue-2832463808
If that is understandable i would take it as comfit message.

Reply:
Things seem to become more clear now - at least this question pops up now:
Would it make sense, or be possible at all, for this test to _notify PDE_ about this "manual" removal of classpath entries before launching UpdateClasspathJob, to avoid the inconsistent knowledge of "old classpath"?

Reply:
I think we can simply ignore any exceptions here, if it does not work PDE can not do much here and likely something will happen short after that refresh the classpath and run a new `UpdateClasspath` job. If not the worst that can happen is that we can't update the source location.

Reply:
ok, done